framework/src/ble_manager/ble_manager_api.c

```diff
@@ -0,0 +1,343 @@
+/****************************************************************************
+ *
+ * Copyright 2021 Samsung Electronics All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+#include <tinyara/config.h>
+
+#include <stdio.h>
+#include <stdint.h>
+#include <ble_manager/ble_manager.h>
+#include "ble_manager_event.h"
+#include "ble_manager_msghandler.h"
+
+#define RETURN_RESULT(res, msg)      \
+	do {                             \
+		if (res < 0) {               \
+			return BLE_MANAGER_FAIL; \
+		} else {                     \
+			return msg.result;       \
+		}                            \
+	} while (0)
+
+ble_result ble_manager_init(ble_client_init_config *client_config, ble_server_init_config *server_config)
+{
+	if (client_config == NULL && server_config == NULL) {
+		BLE_ERR_RECORD(BLE_MANAGER_INVALID_ARGS);
+		return BLE_MANAGER_INVALID_ARGS;
+	}
+
+	blemgr_msg_params param = { 2, {(void *)client_config, (void *)server_config} };
+	blemgr_msg_s msg = {BLE_EVT_CMD_INIT, BLE_MANAGER_FAIL, (void *)(&param), NULL};
+	int res = blemgr_post_message(&msg);
+
+	RETURN_RESULT(res, msg);
+}
+
+ble_result ble_manager_deinit(void)
```

Comment:
I think all apis here has just one fixed error reason, that is BLE_MANAGER_FAIL, you should use more reasons of error.
Also it seems that fail reason decided caller, I mean api in here. WHY???

Reply:
The "res" has only the result of posting message.
So the user check "res" first, and after that, "msg.result" would be the real reason of proccessing.
In conclusion, both of "res" and "msg.result" should be checked for the result